src/libraries/System.Linq.Expressions/src/System/Dynamic/Utils/DelegateHelpers.cs

```diff
@@ -3,28 +3,38 @@
 
 using System.Diagnostics.CodeAnalysis;
 using System.Reflection;
-
-#if !FEATURE_DYNAMIC_DELEGATE
 using System.Reflection.Emit;
 using System.Text;
 using System.Threading;
-#endif
 
 namespace System.Dynamic.Utils
 {
     internal static class DelegateHelpers
     {
-        internal static Delegate CreateObjectArrayDelegate(Type delegateType, Func<object?[], object?> handler)
+        // This can be flipped to true using feature switches at publishing time
+        internal static bool CanEmitObjectArrayDelegate => string.Empty != null; // HACK: complex enough so that ILLink doesn't constprop at library build
```

Comment:
@marek-safar @vitek-karas I had to hack around this like this because this:

* Should be a constant `true`,
* Unless overriden by a feature switch at publish time.

ILLink is propagating the constant true _at library build time_ and I can't figure out how else to prevent it from doing that during library build. It would be fine to constprop at app publish time.

Reply:
Alternatively you could disable constprop during libraries build

Reply:
If this is driven by a feature switch, then it should follow the feature switch pattern - that is ask for the value from `AppContext`. The link substitutions may force it to a const value, but that should only be done on app publish time, not in library build.

Reply:
This should not be a user visible setting - JIT-based CoreCLR flavor will probably never implement the API flipping this switch is going to call; we might want to implement that API on Mono, but it's not implemented right now. This should stay `#define`-like - it should not be selected by a user; it should be selected by us.

I'll see how much other work constprop is doing during library build for this library and if it's okay to fully disable.

Reply:
I made this into constants and disabled ipconsprop instead.

Disabling ipconstprop made the size worse by 512 bytes (less than 0.1%) for non-iDevice workloads. But having an actual constant there will result in better trimming at publish time.

The iOS version size got better by 512 bytes on the other hand - we won't need to switch this one between Ref.Emit and interpreter so it can keep the constprop and making these into constants allowed a tiny bit of code to be elimintated.

Reply:
Yes, I think you summed it up well. This pull request is basically just changing `#define`s into feature switches that are (at least initially) not exposed in user-friendly ways. The main reason is that NativeAOT needs to compile with a different set of defines. There is also an iDevice-specific substitution file that emulates the preexisting behavior of this being a compile time define for those platforms.

In the future, we could entertain an option to expose part of this as a public thing - "always use interpreter" could become a public feature switch that would have an appropriate AppContext value (to emulate when not trimming) and set the feature switch to remove expression compilation support when actually trimming. That's not in scope of this pull request but could be a follow on work item.

This pull request introduces 3 feature switches. One makes some sense to potentially expose publicly in the future (but it would be opt in, not default with trimming). one makes some marginal sense to expose publicly but is probably not worth it, and one is internal-only that we should ideally enable whenever ref-emit is fully disabled in the underlying runtime (once Mono supports that fallback API) - it doesn't make sense to individually expose that one.

Reply:
> One thing I don't understand: This behavior effectively requires trimming to happen in order for the code to be runnable on platforms which require some of the other implementations. How is that supposed to work with debug builds. If I remember correctly even iOS builds don't run trimming for debug builds (they do run linker, but in "copy-only" mode, which will not perform substitutions).

Sorry - I'm probably slow here - how will this work if trimming with substitutions is NOT performed on the code?

Reply:
The substitution happens at repo build time here. There's a substitution file that applies at repo build for iDevices. Ipconstprop is also enabled there. There was no publish time configurability around this for iDevices and there won't be after this pull request.